[Merged by Bors] - Reduce power usage with configurable event loop

aevyrie opened the pull request:
# Objective

- Reduce power usage for games when not focused.
- Reduce power usage to ~0 when a desktop application is minimized (opt-in).
- Reduce power usage when focused, only updating on a `winit` event, or the user sends a redraw request. (opt-in)

https://user-images.githubusercontent.com/2632925/156904387-ec47d7de-7f06-4c6f-8aaf-1e952c1153a2.mp4

Note resource usage in the Task Manager in the above video.

## Solution

- Added a type `UpdateMode` that allows users to specify how the winit event loop is updated, without exposing winit types.
- Added two fields to `WinitConfig`, both with the `UpdateMode` type. One configures how the application updates when focused, and the other configures how the application behaves when it is not focused. Users can modify this resource manually to set the type of event loop control flow they want.
- For convenience, two functions were added to `WinitConfig`, that provide reasonable presets: `game()` (default) and `desktop_app()`.
    - The `game()` preset, which is used by default, is unchanged from current behavior with one exception: when the app is out of focus the app updates at a minimum of 10fps, or every time a winit event is received. This has a huge positive impact on power use and responsiveness on my machine, which will otherwise continue running the app at many hundreds of fps when out of focus or minimized.
    - The `desktop_app()` preset is fully reactive, only updating when user input (winit event) is supplied or a `RedrawRequest` event is sent. When the app is out of focus, it only updates on `Window` events - i.e. any winit event that directly interacts with the window. What this means in practice is that the app uses *zero* resources when minimized or not interacted with, but still updates fluidly when the app is out of focus and the user mouses over the application.
- Added a `RedrawRequest` event so users can force an update even if there are no events. This is useful in an application when you want to, say, run an animation even when the user isn't providing input.
- Added an example `low_power` to demonstrate these changes

## Usage

Configuring the event loop:
```rs
use bevy::winit::{WinitConfig};
// ...
.insert_resource(WinitConfig::desktop_app()) // preset
// or
.insert_resource(WinitConfig::game()) // preset
// or
.insert_resource(WinitConfig{ .. }) // manual
```

Requesting a redraw:
```rs
use bevy::window::RequestRedraw;
// ...
fn request_redraw(mut event: EventWriter<RequestRedraw>) {
    event.send(RequestRedraw);
}
```

## Other details

- Because we have a single event loop for multiple windows, every time I've mentioned "focused" above, I more precisely mean, "if at least one bevy window is focused".
- Due to a platform bug in winit (https://github.com/rust-windowing/winit/issues/1619), we can't simply use `Window::request_redraw()`. As a workaround, this PR will temporarily set the window mode to `Poll` when a redraw is requested. This is then reset to the user's `WinitConfig` setting on the next frame.

alice-i-cecile commented on examples/window/low_power.rs:
```diff
@@ -0,0 +1,69 @@
+    winit::{ControlFlow, WinitConfig},
+};
+
+/// This example illustrates how to run in low power mode, useful for making desktop applications.
```
How does this strategy interact with async tasks, like e.g. slicing a 3D print, that you might want to perform while the app is minimized?

aevyrie replied:
I assume that if work is spawned to a task pool, this won't prevent progress assuming the task pool is also being polled from the other thread. If you really need to keep running the world, you can just send a redraw request event.

alice-i-cecile replied:
I agree with you, but we should test and demonstrate this in the example :)

aevyrie replied:
Hmm, how do you think we could test? Spawn a task set on a timer for 5 sec-> minimize for 5 secs -> restore window and see if the task completes immediately.

alice-i-cecile replied:
> Hmm, how do you think we could test? Spawn a task set on a timer for 5 sec-> minimize for 5 secs -> restore window and see if the task completes immediately.

Yep that sounds perfect!

aevyrie commented on examples/window/low_power.rs:
```diff
@@ -0,0 +1,69 @@
+#[derive(Component)]
+struct Rotator;
+
+/// set up a simple 3D scene
```
Any specific suggestions? I didn't want to distract from what was important in the example.

alice-i-cecile replied:
I was thinking either a) spawn like 10k cubes or b) do some fake work using sleep.

aevyrie replied:
What is the goal here, to show how much power usage drops?

alice-i-cecile replied:
Yeah; we want to be able to let users actually verify that this works on their machine. Going from 1% draw to 0.1% will be much harder to spot.

alice-i-cecile replied:
This is partly for convincing users that it works, but also to allow us to catch weird platform-specific bugs.

aevyrie replied:
I think just disabling vsync will do the trick, that will bog down CPU/GPU regardless of how powerful their PC is - it will basically just run as fast as it can.

alice-i-cecile replied:
Perfect, that's nice and easy.

IceSentry replied:
Doesn't work on g-sync monitors. Mine will stay at 144hz even with v-sync off.

aevyrie replied:
> Doesn't work on g-sync monitors. Mine will stay at 144hz even with v-sync off.

Have you tested this? From what I understand, using immediate mode will prevent the main loop from waiting on sync, regardless of which frames are presented on screen - similar to mailbox mode.

In other words, you might see tear-free rendering capped at your monitor's refresh rate, but the bevy app will still be running as fast as it can in the background. I can only test on a freesync monitor.

IceSentry replied:
What I meant is that disabling vsync doesn't actually unlock the framerate on a g-sync monitor. At least, not consistently, if I move the window between a g-sync and non g-sync monitor it can sometimes actually unlock it. It's really weird, but @cwfitzgerald confirmed that g-sync is weird in this context with wgpu, My point is mostly that it won't change the cpu usage.

IceSentry replied:
https://discord.com/channels/691052431525675048/866787577687310356/927054585363853353 here's a link to a discord discussion about this

aevyrie replied:
I revived that conversation and there weren't any conclusive answers. Could you test the `low_power` example in this branch on the gsync monitor, and report back with resource usage between the three modes?

alice-i-cecile reviewed (changes requested):
The feature is very useful, the strategy checks out and the code quality and docs are good. All I want is a more advanced example and I'll be happy with this.

HackerFoo commented:
I've been testing a similar change, where the app can dynamically control the update interval: https://github.com/HackerFoo/bevy/commit/653b825993703dd356be368b4865680c15821459

I've been testing this for a while on iOS, but I'm still not quite happy with how it works, because there is a pause when resuming the app (between `NewEvents` and the next event) depending on the interval, which should not be there. I think it's a bug in `winit` on iOS, but I haven't been able to fix it yet.

aevyrie commented:
@alice-i-cecile I've updated the example, squashed all the bugs I could find, and tested async tasks using the `async_compute` example. Async tasks worked nicely - spawned tasks ran in the background, and as soon as the application was interacted with, all the results showed up instantly.

aevyrie commented:
@alice-i-cecile Okay, now it's *really* ready for review. 😆 I found a few more bugs and tried to make the example much clearer.

Some interesting notes: 

* Running in `Wait` mode and sending redraw requests every frame appears to be just as performant as using `Poll`. In my CPU-bound example, I'm hitting ~540fps in both scenarios. This suggests we could consider switching to `Wait` mode by default, this would make it easier to have control over the winit event loop, so we can do fancier things for choosing how frequently to update the event loop - whether that's for power use, latency, @HackerFoo's dynamic update interval, or other shenanigans.
* Redraw requests don't build up, so there's no risk of "overusing" the feature. This is useful for composability: a system can submit a redraw request and the app will update to the next frame regardless of what other systems are doing, there is no performance impact from having various systems cumulatively submitting 1,000 redraw requests. Essentially, `Wait` mode will only ever pause if there are no winit or redraw events.
* Once paused, there is no way to make the app start updating when using `Wait`, unless a winit event comes in. There are two solutions I can see: 1) add a custom winit event on construction of the eventloop, and pass the proxy into the app world. An external thread can then send an event to trigger the winit loop to continue. This will require more work. 2) Use `WaitUntil(Instant)` instead of `Wait`. This will ensure the app will eventually wake up once the timeout is reached. It's less flexible than (1), but easier to use and already possible with this PR.

superdump commented:
Tested on macOS on an M1 Max:
<img width="106" alt="Screenshot 2022-02-25 at 23 07 21" src="https://user-images.githubusercontent.com/302146/155809769-c006733a-3cf3-4d97-be3a-83d5e842bdb7.png">
The drop in the CPU/GPU usage to much lower levels is when Wait mode is used. The higher parts are with WaitAndRedraw or Poll. Very nice!

aevyrie commented:
I have noticed what I would consider a bug. When the window is in the background, it still updates with events when it shouldn't. Strangely, in some cases it _does_ work as expect - if I select the Task Manger, the bevy app will stop updating unless the mouse is actually hovering over the background bevy window. My guess is this behavior may be platform dependent.

I'm tempted to special-case this, such that while the window does not have focus, the only events that trigger an update are mouse window events (or explicit redraw requests).

alice-i-cecile commented:
> I'm tempted to special-case this, such that while the window does not have focus, the only events that trigger an update are mouse window events (or explicit redraw requests).

I don't mind this, provided we leave good comments in the code.

mockersf commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -7,6 +7,7 @@ use bevy_input::{
     mouse::{MouseButtonInput, MouseMotion, MouseScrollUnit, MouseWheel},
     touch::TouchInput,
 };
+pub use winit::event_loop::ControlFlow;
```
This is the only type reexported from winit in Bevy. I would prefer if it was mapped to a Bevy struct.

mockersf replied:
```suggestion
use winit::event_loop::ControlFlow;
```

mockersf commented:
> I have noticed what I would consider a bug. When the window is in the background, it still updates with events when it shouldn't. Strangely, in some cases it _does_ work as expect - if I select the Task Manger, the bevy app will stop updating unless the mouse is actually hovering over the background bevy window. My guess is this behavior may be platform dependent.

I've noticed the same thing on macOS, I don't think it's a bug, the OS is sending redraw requests.

aevyrie commented:
Here's a bit more context as to why I opened this PR, and some of the detours I took to get here, please excuse the long response. 🙂 

> I've noticed the same thing on macOS, I don't think it's a bug, the OS is sending redraw requests.

That's fair - by definition it's responding to OS events and the behavior is _technically_ correct. Regardless, I don't think that's appropriate behavior for the problems this feature is attempting to solve. To make Bevy useful for application development (or power-friendly games!), we need a way to:

- Run in a low power state when the application isn't in use, out of the box

and/or

- Give developers the tools to manually adjust the event loop to do it themselves without forking `bevy_winit` and `bevy_window`

We're currently testing an application with users, and this has been a high priority issue. Basically "why does this app make my computer slow even when it's in the background?". Obviously, we can just fork Bevy, but I'm hoping we can produce at least a partial solution to the problem in this PR. I think this will be a common issue for anyone developing applications with Bevy, @HackerFoo's comments w.r.t. `Noumenal` here and on discord appear to support this conclusion. In other words, the current behavior is technically correct in `Wait` mode, but that's not the goal - the goal is reasonable and unsurprising resource usage for applications.

One of the solutions I investigated to solve this without forking bevy, was to take `bevy_framepace` and set the framerate limit absurdly low when the window was unfocused. Unfortunately, this had some unintended side effects:

- When restoring the minimized application, it would hang for the duration of the frame limit before showing, because it is a blocking timer.
- Because of this first issue, you can't just pause the app until it is restored (un-minimized), because it will be permanently blocked, so you need to update at _some_ interval.
- When the app updated in the background, it would still render a frame and run all systems, which would cause large, repeated spikes in resource usage.

In many ways, this is relevant to your comment about not re-exporting a `winit` type and exposing it as part of the Bevy API. Maybe we could provide two default update modes, `Continuos` and `Reactive`.

- `Continuous` would be identical to current bevy behavior. In the future we could possibly expand this with a rate limit.
- `Reactive` would behave like current `Wait` when focused but would only update on events inside the window's extents when unfocused (e.g. `CursorMoved`). It might be worth using `WaitUntil` if the user wants to specify a minimum update interval.

mockersf commented:
> I'm starting to think we should instead expose some bevy setting for update modes. In many ways, this is relevant to your comment about not re-exporting a `winit` type as part of the Bevy API. Maybe we could provide two default update modes, `Continuos` and `Reactive`.

Sounds good to me! Using an enum with new words let us choose their meaning rather than reusing words with existing meaning

aevyrie commented:
I've implemented the changes laid out in my comment above. Please see the new video, and rewritten first post for details.

Weasy666 reviewed (commented):
If i noticed this correctly in Discord, then we want to replace `.unwrap()`s in Bevy with `.expect("Meaning full panic message")` to get more meaningful info/messages on panics. If that is true, could you replace the "new" `.unwrap()`s in this PR with `.expect()`s?

Rest looks good to me.

superdump commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -227,41 +233,78 @@ pub fn winit_runner(app: App) {
-        .map_or(false, |config| config.return_from_run);
-
-    let mut active = true;
+        .unwrap()
```
Corresponding to the use of `.resource()`
```suggestion
```

superdump commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -227,41 +233,78 @@ pub fn winit_runner(app: App) {
-
         match event {
+            event::Event::NewEvents(start) => {
+                let winit_state = app.world.get_resource::<WinitPersistentState>().unwrap();
+                let winit_config = app.world.get_resource::<WinitConfig>().unwrap();
+                let windows = app.world.get_resource::<Windows>().unwrap();
```
```suggestion
                let winit_state = app.world.resource::<WinitPersistentState>();
                let winit_config = app.world.resource::<WinitConfig>();
                let windows = app.world.resource::<Windows>();
```

superdump commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -227,41 +233,78 @@ pub fn winit_runner(app: App) {
+                let winit_config = app.world.get_resource::<WinitConfig>().unwrap();
+                let windows = app.world.get_resource::<Windows>().unwrap();
+                let focused = windows.iter().any(|w| w.is_focused());
+                // Check if either the `WaitUntil` timeout was reached, or
```
Incomplete sentence

superdump commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -227,41 +233,78 @@ pub fn winit_runner(app: App) {
+                };
+                let mut winit_state = app
+                    .world
+                    .get_resource_mut::<WinitPersistentState>()
+                    .unwrap();
```
```suggestion
                    .resource_mut::<WinitPersistentState>();
```

superdump commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -227,41 +233,78 @@ pub fn winit_runner(app: App) {
+                let now = Instant::now();
+                let manual_timeout_reached = match winit_config.update_mode(focused) {
+                    UpdateMode::Continuous => false,
+                    UpdateMode::Reactive { max_wait } => {
+                        now.duration_since(winit_state.last_update) >= *max_wait
+                    }
+                    UpdateMode::ReactiveLowPower { max_wait } => {
+                        now.duration_since(winit_state.last_update) >= *max_wait
+                    }
```
These two cases are identical so they can be collapsed (maybe `cargo fmt` after):
```suggestion
                    UpdateMode::Reactive { max_wait } | UpdateMode::ReactiveLowPower { max_wait } => {
                        now.duration_since(winit_state.last_update) >= *max_wait
                    }
```

superdump commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -497,25 +545,87 @@ pub fn winit_runner_with(mut app: App) {
             }
             event::Event::Resumed => {
-                active = true;
+                app.world
+                    .get_resource_mut::<WinitPersistentState>()
+                    .unwrap()
```
```suggestion
                app.world
                    .resource_mut::<WinitPersistentState>()
```

superdump commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -497,25 +545,87 @@ pub fn winit_runner_with(mut app: App) {
+                    false
+                };
+                if update {
+                    app.world
+                        .get_resource_mut::<WinitPersistentState>()
+                        .unwrap()
```
```suggestion
                    app.world
                        .resource_mut::<WinitPersistentState>()
```

superdump commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -497,25 +545,87 @@ pub fn winit_runner_with(mut app: App) {
+                    let windows = app.world.get_resource::<Windows>().unwrap();
+                    let focused = windows.iter().any(|w| w.is_focused());
+                    match winit_config.update_mode(focused) {
+                        UpdateMode::Continuous => true,
+                        UpdateMode::Reactive { .. } => true,
```
```suggestion
                        UpdateMode::Continuous | UpdateMode::Reactive { .. } => true,
```

superdump commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -497,25 +545,87 @@ pub fn winit_runner_with(mut app: App) {
+                    use UpdateMode::*;
+                    *control_flow = match winit_config.update_mode(focused) {
+                        Continuous => ControlFlow::Poll,
+                        Reactive { max_wait } => ControlFlow::WaitUntil(now + *max_wait),
+                        ReactiveLowPower { max_wait } => ControlFlow::WaitUntil(now + *max_wait),
```
```suggestion
                        Reactive { max_wait } | ReactiveLowPower { max_wait } => {
                            ControlFlow::WaitUntil(now + *max_wait)
                        }
```

superdump commented on crates/bevy_winit/src/winit_config.rs:
```diff
@@ -12,4 +14,76 @@ pub struct WinitConfig {
+            ..Default::default()
+        }
+    }
+    pub fn update_mode(&self, focused: bool) -> &UpdateMode {
```
```suggestion

    /// Gets the configured `UpdateMode` depending on whether the window is focused or not
    pub fn update_mode(&self, focused: bool) -> &UpdateMode {
```

superdump commented on crates/bevy_winit/src/winit_config.rs:
```diff
@@ -12,4 +14,76 @@ pub struct WinitConfig {
+            return_from_run: false,
+            focused_mode: UpdateMode::Continuous,
+            unfocused_mode: UpdateMode::ReactiveLowPower {
+                max_wait: Duration::from_millis(100),
```
I don't know if default impls can be documented but it would be cool if it were clear somewhere that by default the frame rate when unfocused is limited to 10fps.

superdump commented on crates/bevy_winit/src/winit_config.rs:
```diff
@@ -12,4 +14,76 @@ pub struct WinitConfig {
+    /// ## Note
+    ///
+    /// Once the app has executed all bevy systems and reaches the end of the event loop, there is
+    /// no way to force the app wake and update again, unless a winit event is received or the time
```
Maybe people don't know what a winit event is, so perhaps an example like an input event or whatever is correct?

superdump reviewed (changes requested):
Minor suggestions. This is a fantastic PR, well done!

superdump commented:
> If i noticed this correctly in Discord, then we want to replace `.unwrap()`s in Bevy with `.expect("Meaning full panic message")` to get more meaningful info/messages on panics. If that is true, could you replace the "new" `.unwrap()`s in this PR with `.expect()`s?

I was going to say the same but there are now infallible `.resource()` and `.resource_mut()` that should be used in cases where the resource will always exist (barring that someone removes the resource...) and they panic internally if the resource is not found. I made suggestions to fix all of those cases above.

aevyrie commented:
Really excellent feedback. Thanks @superdump!

Weasy666 commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -288,6 +325,11 @@ pub fn winit_runner_with(mut app: App) {
 
+                world
+                    .get_resource_mut::<WinitPersistentState>()
+                    .unwrap()
```
Did you miss this sneaky one `unwrap()`? 🙂

aevyrie replied:
Unfortunately `world.cell()` does not implement the infallible `.resource_mut()`, although I should change that to an `expect`. Thanks! 😄

aevyrie replied:
This section is filled with pre-existing `unwrap`s. It might be better to implement the infallible functions on `world.cell()` instead of filling this section with repeated `.expect("unable to access resource")`.

aevyrie replied:
@Weasy666 I've created #4104 to address this. Good spot!

superdump commented on examples/window/low_power.rs:
```diff
@@ -70,7 +70,7 @@ fn update_winit(
 /// Switch between update modes when the mouse is clicked.
 fn cycle_modes(mut mode: ResMut<ExampleMode>, mouse_button_input: Res<Input<MouseButton>>) {
-    if mouse_button_input.just_pressed(MouseButton::Left) {
+    if mouse_button_input.just_pressed(MouseButton::Right) {
```
This makes it a little more difficult to use on laptops perhaps.

kirusfg replied:
Well, I am speculating people using laptops won't find it difficult because they are used to it, there is always either a physical rmb or a two-finger tap. Curious to know why change it to right mouse button at all.

aevyrie replied:
I changed it because I found when trying out the demonstrated features - namely while defocusing and refocusing the window - I would inadvertently switch modes as I clicked on the window to bring it into focus.

I could change to spacebar to make it easier to switch intentionally.

kirusfg replied:
Oh, it is well-justified then. 
Anything else than LMB is fine.

mockersf commented on examples/app/return_after_run.rs:
```diff
@@ -5,6 +5,7 @@ fn main() {
     App::new()
         .insert_resource(WinitConfig {
             return_from_run: true,
+            ..Default::default()
```
examples should now use the new `..default()`

mockersf commented on examples/app/return_after_run.rs:
```diff
@@ -14,6 +15,7 @@ fn main() {
     App::new()
         .insert_resource(WinitConfig {
             return_from_run: true,
+            ..Default::default()
```
```suggestion
            ..default()
```

mockersf commented on examples/window/low_power.rs:
```diff
@@ -0,0 +1,221 @@
+            unfocused_mode: bevy::winit::UpdateMode::ReactiveLowPower {
+                max_wait: Duration::from_millis(10),
+            },
+            ..Default::default()
```
```suggestion
            ..default()
```

mockersf commented on examples/window/low_power.rs:
```diff
@@ -0,0 +1,221 @@
+        // Turn off vsync to maximize CPU/GPU usage
+        .insert_resource(WindowDescriptor {
+            present_mode: PresentMode::Immediate,
+            ..Default::default()
```
```suggestion
            ..default()
```

mockersf commented on examples/window/low_power.rs:
```diff
@@ -0,0 +1,221 @@
+            .spawn_bundle(PbrBundle {
+                mesh: meshes.add(Mesh::from(shape::Cube { size: 0.5 })),
+                material: materials.add(Color::rgb(0.8, 0.7, 0.6).into()),
+                ..Default::default()
```
```suggestion
            ..default()
```

mockersf commented on examples/window/low_power.rs:
```diff
@@ -0,0 +1,221 @@
+            point_light: PointLight {
+                intensity: 1500.0,
+                shadows_enabled: true,
+                ..Default::default()
```
```suggestion
            ..default()
```

mockersf commented on examples/window/low_power.rs:
```diff
@@ -0,0 +1,221 @@
+                ..Default::default()
+            },
+            transform: Transform::from_xyz(4.0, 8.0, 4.0),
+            ..Default::default()
```
```suggestion
            ..default()
```

mockersf replied:
well you get the idea for all of them :D

mockersf commented on crates/bevy_winit/src/winit_config.rs:
```diff
@@ -1,15 +1,92 @@
+        WinitConfig {
+            return_from_run: false,
+            focused_mode: UpdateMode::Continuous,
+            unfocused_mode: UpdateMode::ReactiveLowPower {
+                max_wait: Duration::from_millis(100),
```
I'm not sure I like having this as a default, it feels very arbitrary. I'm not sure I don't like either though... Any idea on what are the defaults for other engine?

aevyrie replied:
Honestly, I just set it to something that looked similar to other games that slow down when not focused, and had a significant reduction in resource usage. I feel the same way about it.

aevyrie replied:
Maybe we just leave this unresolved and get cart's input.

cart replied:
Oops left a comment [here](https://github.com/bevyengine/bevy/pull/3974#discussion_r821161493) before I noticed this open thread.

mockersf commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -17,16 +18,19 @@ use bevy_ecs::{
     dpi::PhysicalPosition,
-    event::{self, DeviceEvent, Event, WindowEvent},
-    event_loop::{ControlFlow, EventLoop, EventLoopWindowTarget},
+    event::{self, DeviceEvent, Event, StartCause, WindowEvent},
+    event_loop::{EventLoop, EventLoopWindowTarget},
```
```suggestion
    event::{self, DeviceEvent, Event, StartCause, WindowEvent},
    event_loop::{ControlFlow, EventLoop, EventLoopWindowTarget},
```

mockersf commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -37,6 +41,8 @@ pub struct WinitPlugin;
 impl Plugin for WinitPlugin {
     fn build(&self, app: &mut App) {
         app.init_non_send_resource::<WinitWindows>()
+            .init_resource::<WinitPersistentState>()
```
Is this needed as a resource? It seems it's never used outside of the event loop, couldn't it be a captured variable there?

alice-i-cecile replied:
End users may want to configure this at run-time, so I like it as a resource.

mockersf replied:
This struct is only an internal state, it should not be touched by end users. And it's private, so not usable anyway

aevyrie replied:
What is the proper way to persist state here without a resource?

aevyrie commented:
This now has all feedback incorporated from all reviewers. 😅

I'm going to stop all changes here until cart's review.

cart commented on crates/bevy_winit/src/winit_config.rs:
```diff
@@ -1,15 +1,92 @@
+        WinitConfig {
+            return_from_run: false,
+            focused_mode: UpdateMode::Continuous,
+            unfocused_mode: UpdateMode::ReactiveLowPower {
```
I don't think falling back to 10 FPS when unfocused is a good default for games:
* Games build logic based on fixed timesteps. For things like physics, ticking at 10 FPS increases the likelihood of "tunneling" and things of that nature. Obviously games should try to protect against tick-time-based discrepancies in behavior because perf issues do happen, but completely insulating against sustained low-tick-rates will be challenging for some types of games.  
* For online games that send and receive network events from inside a game loop, these should still be processed at "reasonable" intervals to encourage proper state syncs. What constitutes "reasonable" is game-specific. I think the only safe assumption to make by default is "developer configured fps limits for active gameplay".
* A common case for "unfocusing games" will be "playing a windowed game and alt-tabbing to another window". In this case, an unfocused game might still be visible and the "perceived quality" of the game will go down if it starts "lagging" when alt-tabbed.

I think throttling _rendering_ (not ticks) by default when unfocused is a slightly more compelling argument. But until we decouple the two (ex: the "actual pipelining" impl would do this), I think we should default to `unfocused_mode = UpdateMode::Continuous` for games.

cart reviewed (commented):
Great work! I think this is generally a good direction to take this. Might need a few changes here and there when "actual pipelining" gets merged, but it largely seems compatible with that.
Just a couple of comments to resolve, then I think this is good to go!

cart commented:
bors r+

xeonmc commented:
>> Due to a platform bug in winit (https://github.com/rust-windowing/winit/issues/1619), we can't simply use Window::request_redraw(). As a workaround, this PR will temporarily set the window mode to Poll when a redraw is requested. This is then reset to the user's WinitConfig setting on the next frame.

Would this mean that the timestamping requirement I described in #4502 would not be achievable with Bevy or winit in general?

aevyrie commented:
> >> Due to a platform bug in winit (https://github.com/rust-windowing/winit/issues/1619), we can't simply use Window::request_redraw(). As a workaround, this PR will temporarily set the window mode to Poll when a redraw is requested. This is then reset to the user's WinitConfig setting on the next frame.
> 
> Would this mean that the timestamping requirement I described in #4502 would not be achievable with Bevy or winit in general?

I'm not sure, I don't understand your request. I'll reply in that thread.
